### PR TITLE
Feature: custom virtual environment directories

### PR DESCRIPTION
**Pull request by ghost**
Currently, users are allowed to define virtual environments in `settings.BASE_VENV_PATH` only, because that's the only place Tower looks for virtual environments. This feature allows users
to custom define the directory paths, using API or UI, to look for virtual environments. Tower aggregates virtual environments from all these paths, except environments with special name `awx`.

Signed-off-by: Vismay Golwala <vgolwala@redhat.com>

##### SUMMARY

The current feature adds a new system setting called ```CUSTOM_VENV_PATHS```. This setting is a list of string paths, which tells Tower places to look for virtual environments. While fetching custom virtual environments, Tower aggregates environments from all these paths. However, environments named ```awx``` will not be fetched as it is a special name.

The system setting is modifiable using general HTTP requests to the API, as well as, under system settings in the UI. It is enforced that the path specified must be a valid file system path. Also, the resultant path is normalized and then duplicates are eliminated. So a request like:

```HTTP PATCH /api/v2/settings/system {'CUSTOM_VENV_PATHS': ['/opt/', '/opt', '/opt']``` 

will set

``` CUSTOM_VENV_PATHS: ['/opt']```

because ```/opt/``` is normalized to ```/opt``` and then the duplicates are removed.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API
 - UI

##### AWX VERSION

```
awx: 3.0.1
```

##### ADDITIONAL INFORMATION
Refer #3126

### CODE REVIEW

File: awx/conf/fields.py [hunk 1]
```diff
@@ -96,6 +100,17 @@ def to_internal_value(self, data):
         self.fail('type_error', input_type=type(data))
 
 
+class StringListPathField(StringListField):
+
+    def __init__(self, **kwargs):
+        super(StringListPathField, self).__init__(**kwargs)
+        self.validators.append(lambda sp: list(map(validate_system_path, sp)))
```
Comment (ryanpetrello) [inline awx/conf/fields.py]: These nested functions are a little hard to parse (at least for my brain 😄), would a list comprehension be clearer to the reader?

```python
self.validators.append(lambda paths: [validate_system_path(path) for path in paths])
```

File: awx/conf/fields.py [hunk 2]
```diff
@@ -96,6 +100,17 @@ def to_internal_value(self, data):
...
+
+    def to_representation(self, value):
+        # Normalize each path value and filter potential duplicates
+        return super(StringListField, self).to_representation(list(set(map(os.path.normpath, value))))
```
Comment (ryanpetrello) [inline awx/conf/fields.py]: Same suggestion here, let's make this a list comprehension, or just a loop so it's easier to read.

File: awx/main/tests/unit/test_validators.py [hunk 3]
```diff
@@ -157,3 +158,13 @@ def test_valid_vars(var_str):
 def test_invalid_vars(var_str):
     with pytest.raises(RestValidationError):
         vars_validate_or_raise(var_str)
+
+
+def test_valid_system_path():
+    validate_system_path("/var")
```
Comment (ryanpetrello) [inline awx/main/tests/unit/test_validators.py]: This probably needs an assertion to be a meaningful test.  Otherwise, it's just code that runs without any side effects.  Something like:

```python
assert validate_system_path("/var") is True
```

File: awx/main/tests/unit/test_validators.py [hunk 4]
```diff
@@ -157,3 +158,13 @@ def test_valid_vars(var_str):
...
+
+
+def test_invalid_system_path():
+    with pytest.raises(ValidationError) as ve:
+        validate_system_path("/foo/bar")
+    assert ve.match('/foo/bar is not a valid path choice.')
```
Comment (ryanpetrello) [inline awx/main/tests/unit/test_validators.py]: 👍

**Review by ryanpetrello — COMMENTED** [review]
Excellent work on this, it is looking good 😄 

Let's add some API tests for the new settings validation to verify that it works.  In terms of where to put them, this is a good place to start:

https://github.com/ansible/tower/blob/devel/awx/conf/tests/functional/test_api.py

It would be good to write some tests that:

- Provide a valid list of custom paths via `HTTP PATCH /api/v2/settings/system/ {"CUSTOM_VENV_PATHS": [...]}` and assert that we got back a reasonable response code for valid paths.
- Provide a variety of _bad_ inputs (lists with missing paths, lists with numbers or dicts in them, inputs that aren't lists _at all_ etc...) and make sure we get back a reasonable response code (maybe `HTTP 400`?)  This test would be a good candidate for the usage of `@pytest.mark.parametrize`.

**Comment by ryanpetrello** [comment]
@vismay-golwala as part of this PR, let's *also* update the developer docs here:

https://github.com/ansible/awx/blob/devel/docs/custom_virtualenvs.md

Let's add a new section at the bottom, maybe something like:

## Discovering Custom Virtualenvs in Other Directories
...

File: awx/conf/fields.py (hunk 2 above)
Comment (ryanpetrello) [reply awx/conf/fields.py]: Good idea w/ `normpath` to catch duplicates w/ trailing `/`.

File: awx/main/conf.py [hunk 5]
```diff
@@ -126,6 +126,17 @@ def _load_default_license_from_file():
     category_slug='system',
 )
 
+register(
+    'CUSTOM_VENV_PATHS',
+    field_class=fields.StringListPathField,
+    label=_('Custom virtual environment paths'),
+    help_text=_('Paths where Tower will look for custom virtual environments. '
+                'Enter one path per line.'),
```
Comment (ryanpetrello) [inline awx/main/conf.py]: For clarity, let's say:

```
Paths where Tower will look for custom virtual environments (in addition to /var/lib/awx/venv/).  Enter one path per line.
```

**Comment by vismay-golwala** [comment]
recheck

File: awx/conf/fields.py (hunk 1 above)
Comment (ghost) [reply awx/conf/fields.py]: Ah yes, this is definitely more readable. I will do that!

File: awx/main/tests/unit/utils/test_common.py [hunk 6]
```diff
@@ -178,10 +178,21 @@ def test_get_custom_venv_choices():
...
         os.makedirs(os.path.join(temp_dir, 'bin', 'activate'))
+
+        custom_venv_dir = os.path.join(temp_dir, 'custom')
+        custom_venv_1 = os.path.join(custom_venv_dir, 'venv-1')
+        custom_venv_awx = os.path.join(custom_venv_dir, 'custom', 'awx')
+
+        os.makedirs(os.path.join(custom_venv_1, 'bin', 'activate'))
+        os.makedirs(os.path.join(custom_venv_awx, 'bin', 'activate'))
+
+        settings.CUSTOM_VENV_PATHS = [custom_venv_dir]
```
Comment (ryanpetrello) [inline awx/main/tests/unit/utils/test_common.py]: One problem we regularly have with changing settings in this way is that doing this and _not_ resetting the original value can sometimes cause this state to leak between tests, and can have side effects when *other* tests run.  In really bad situations, it can cause tests failures that **only** occur when running all the tests, but tests pass individually 😱.

There are ways around this, like mocking `settings`, and a variety of decorators and context managers that can automatically _reset_ the originally value after this test runs, but they can be messy and easy to get wrong.

One pattern I really like for situations like this is to add an optional argument e.g.,

```python
def get_custom_venv_choices(custom_paths=None):
    custom_paths = custom_paths or settings.CUSTOM_VENV_PATHS
```

...and then have the *test* just call:

`common.get_custom_venv_choices([custom_venv_dir])`

Comment (ghost) [reply awx/main/tests/unit/utils/test_common.py]: Yes, I believe that's a problem if we directly set the variables. And although, redefining the function is an option here, I am thinking of going with the mock approach. Because it is more elegant that way as we don't have to modify our code pertaining to tests, but instead, we modify tests pertaining to code.

```
with mock.patch('django.conf.settings.CUSTOM_VENV_PATHS', [custom_venv_dir]):
            assert sorted(common.get_custom_venv_choices()) == [
            ....
```

Comment (ryanpetrello) [reply awx/main/tests/unit/utils/test_common.py]: > Because it is more elegant that way as we don't have to modify our code pertaining to tests, but instead, we modify tests pertaining to code.

There _is_ a distinction between modifying code to accommodate tests, and writing code that is _easily testable_.  The pattern I'm referring to above is known as dependency injection, and you'll see it used a lot throughout the unit tests in our codebase as an _alternative_ to mocking.

That said, much of this comes down to personal preference, and you'll find that every developer you talk to has a different opinion on when it is and isn't appropriate to use a mock 😄.

File: awx/main/validators.py [hunk 7]
```diff
@@ -197,3 +198,12 @@ def vars_validate_or_raise(vars_str):
...
         raise RestValidationError(str(e))
+
+
+def validate_system_path(value):
+    """
+    Validate that string passed as a path actually exists in the
+    file system. If not, raise a ValidationError.
+    """
+    if not os.path.exists(value):
+        raise ValidationError(_('{} is not a valid path choice').format(value))
```
Comment (wenottingham) [inline awx/main/validators.py]: Do we want to do some more checks here that the selected path isn't obviously terrible (and/or "immediately blocked by bwrap")?

Comment (ryanpetrello) [reply awx/main/validators.py]: Eh, I'm not sure we need this.  We already have specific code that explicitly whitelists the virtualenv for the current job run (cc @chrismeyersfsu):

https://github.com/ansible/awx/blob/devel/awx/main/tasks.py#L1001

Comment (wenottingham) [reply awx/main/validators.py]: OK. More concerned about someone entering "CUSTOM_VIRTUALENV_PATH=/root" or someplace else they shouldn't.

Comment (ryanpetrello) [reply awx/main/validators.py]: Meh.  Do dumb things, get dumb prizes.

File: awx/main/tests/unit/utils/test_common.py (hunk 6 above)
Comment (ghost) [reply awx/main/tests/unit/utils/test_common.py]: I didn't know about that, thanks for pointing it out. Then I will go with this approach. :+1:

File: awx/ui/client/src/configuration/forms/system-form/sub-forms/system-misc.form.js [hunk 8]
```diff
@@ -54,6 +54,10 @@ export default ['i18n', function(i18n) {
             REMOTE_HOST_HEADERS: {
                 type: 'textarea',
                 reset: 'REMOTE_HOST_HEADERS'
+            },
+            CUSTOM_VENV_PATHS: {
```
Comment (mabashian) [inline awx/ui/client/src/configuration/forms/system-form/sub-forms/system-misc.form.js:58]: 👍

File: awx/conf/fields.py (hunk 2 above)
Comment (jbradberry) [reply awx/conf/fields.py]: I think we would want a stable ordering of the list, so maybe also sort it?

File: awx/main/tests/unit/utils/test_common.py (hunk 6 above)
Comment (jbradberry) [reply awx/main/tests/unit/utils/test_common.py]: The `django.test` module has the `override_settings` and `modify_settings` decorators that may still work even though we aren't using Django's built-in testing classes here.

Comment (ghost) [reply awx/main/tests/unit/utils/test_common.py]: Thanks @jbradberry. I thought of that at first, but since we are aiming to make it easy for future developers to test their code, I have for now, gone with the dependency injection approach. Let's discuss on this more, if we feel we can make some changes.

**Review by ryanpetrello — APPROVED** [review]
This looks really good to me, great work on this 👍

File: docs/custom_virtualenvs.md [hunk 9]
```diff
@@ -16,29 +16,36 @@ virtualenv; this documentation describes the supported way to do so.
...
-runs.  To choose a custom virtualenv, first create one in `/var/lib/awx/venv`:
+runs.  To choose a custom virtualenv, first we need to create one. Here, we are
+using `/opt/my-envs/` as the directory to hold custom venvs. But you can use any
+other directory and replace `/opt/my-envs/` with that. Let's create the directory
+first if absent:
 
-    $ sudo virtualenv /var/lib/awx/venv/my-custom-venv
+    $ sudo mkdir /opt/my-envs
+
+Now that we have the directory, we can create virtual environment in that using:
```
Comment (ryanpetrello) [inline docs/custom_virtualenvs.md]: ...we can create _a_ virtual environment...

File: docs/custom_virtualenvs.md [hunk 10]
```diff
@@ -60,8 +67,9 @@ index aa8b304..eb05f91 100644
     fi
  
 +requirements_custom:
-+	virtualenv $(VENV_BASE)/my-custom-env
-+	$(VENV_BASE)/my-custom-env/bin/pip install python-memcached psutil
```
Comment (vismay-golwala) [inline docs/custom_virtualenvs.md]: Mistake here, fixing it. Should be `/opt/my-envs/my-custom-env/bin/pip install python-memcached psutil`

File: docs/custom_virtualenvs.md [hunk 11]
```diff
@@ -47,7 +54,7 @@ can be found using `pip freeze`:
 
 One important item to keep in mind is that in a clustered awx installation,
 you need to ensure that the same custom virtualenv exists on _every_ local file
-system at `/var/lib/awx/venv/`.  For container-based deployments, this likely
+system at `/opt/my-envs/`.  For container-based deployments, this likely
 means building these steps into your own custom image building workflow, e.g.,
```
Comment (kdelee) [inline docs/custom_virtualenvs.md:70]: Don't you need to then include adding it to `CUSTOM_VENV_PATHS` ?

Comment (ryanpetrello) [reply docs/custom_virtualenvs.md:70]: Yep, good point.
